packages/vm/src/environment.rs

```diff
@@ -14,6 +14,62 @@ use crate::errors::{VmError, VmResult};
 #[derive(Debug)]
 pub enum Never {}
 
+/** gas config data */
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct GasConfig {
+    /// Gas costs of VM (not Backend) provided functionality
+    /// secp256k1 signature verification cost
+    pub secp256k1_verify_cost: u64,
+    /// secp256k1 public key recovery cost
+    pub secp256k1_recover_pubkey_cost: u64,
+    /// ed25519 signature verification cost
+    pub ed25519_verify_cost: u64,
+    /// ed25519 batch signature verification cost
+    pub ed25519_batch_verify_cost: u64,
+    /// ed25519 batch signature verification cost (single public key)
+    pub ed25519_batch_verify_one_pubkey_cost: u64,
+}
+
+impl GasConfig {
+    // Base crypto-verify gas cost: 1000 Cosmos SDK * 100 CosmWasm factor
+    const BASE_CRYPTO_COST: f32 = 100_000.0;
+
+    // secp256k1 cost factor (reference)
+    const SECP256K1_VERIFY_FACTOR: f32 = 1.; // ~154 us in crypto benchmarks
+
+    // Gas cost factors, relative to secp256k1_verify cost
+    const SECP256K1_RECOVER_PUBKEY_FACTOR: f32 = 162. / 154.; // 162 us / 154 us ~ 1.05
+    const ED25519_VERIFY_FACTOR: f32 = 63. / 154.; // 63 us / 154 us ~ 0.41
+
+    // Gas cost factors, relative to ed25519_verify cost
+    // From https://docs.rs/ed25519-zebra/2.2.0/ed25519_zebra/batch/index.html
+    const ED255219_BATCH_VERIFY_FACTOR: f32 = GasConfig::ED25519_VERIFY_FACTOR / 2.; // 0.41 / 2. ~ 0.21
+    const ED255219_BATCH_VERIFY_ONE_PUBKEY_FACTOR: f32 = GasConfig::ED25519_VERIFY_FACTOR / 4.; // 0.41 / 4. ~ 0.1
+
+    fn calc_crypto_cost(factor: f32) -> u64 {
+        (GasConfig::BASE_CRYPTO_COST * factor).round() as u64
+    }
```

Comment:
Unfortuntely float behaviour is architecture specific and you can even configure rounding behaviour differently in different threads on one CPU without violating the spec. Then there are all those bugs. Can we try to avoid the use of floats completely?

What you can do instead is using rational numbers expressed as `Fraction { nominator, denominator }`. Using those, you can do everything you did above with integer math.

Reply:
Oh, I thought floats were forbidden only in smart contracts code. Will change.